### PR TITLE
LCP: CreateCopy(): fix crash on negative pixel values (fixes #7561)

### LINKED ISSUE #7561: LCP driver failure "double free or corruption (out)" on  CreateCopy or gdal_translate
## Expected behavior and actual behavior.

I am unable to create or save gdal LCP files without a memory error. 

Specifically, I have downloaded a GeoTiff file from the landfire server (https://landfire.gov/) using python landfire api (https://github.com/FireSci/landfire-python). The result is an 8-band GeoTiff with bands as described at https://gdal.org/drivers/raster/lcp.html

Any attempt to CreateCopy() through the python gdal api, to use r.out.gdal in GRASS 8, or to directly call gdal_translate all end in the same error

## Steps to reproduce the problem.

```
gdal_translate -of LCP test.tif test.lcp
```
OR
```
gdal_translate -ot INT16 -of LCP test.tif test.lcp
```

OR
```Python
file = "test.tif"
ds = gdal.Open(os.path.abspath(file))
driver = gdal.GetDriverByName("LCP")
outdata = driver.CreateCopy(os.path.abspath("test.lcp"), ds, 1)
```

All result in

```
double free or corruption (out)
```

And core dump.
## Operating system

Ubuntu 22.04

## GDAL version and provenance

GDAL 3.6.2 from ubuntugis-unstable PPA  But error is same under ubuntugis-stable and distro packages as well.

### PR DESCRIPTION
**Pull request by rouault**

### CODE REVIEW

File: frmts/raw/lcpdataset.cpp [hunk 1]
```diff
@@ -811,43 +811,21 @@ GDALDataset *LCPDataset::Open(GDALOpenInfo *poOpenInfo)
...
-    const int nSpan = static_cast<int>(dfMax);
     GInt16 *panValues =
         static_cast<GInt16 *>(CPLMalloc(sizeof(GInt16) * nXSize));
-    GByte *pabyFlags =
-        static_cast<GByte *>(CPLMalloc(sizeof(GByte) * nSpan + 1));
-    memset(pabyFlags, 0, nSpan + 1);
+    GByte *pabyFlags = static_cast<GByte *>(CPLCalloc(1, 65536));
+
+    /* so that values in [-32768,32767] are mapped to [0,65535] in pabyFlags */
+    constexpr int OFFSET = 32768;
```
Comment (elpaso) [inline frmts/raw/lcpdataset.cpp:828]: does it make sense to use numeric_limits here?

File: frmts/raw/lcpdataset.cpp [hunk 2]
```diff
@@ -856,48 +834,49 @@ CPLErr LCPDataset::ClassifyBandData(GDALRasterBand *poBand,
...
-    // offsets, or making it a 1-based array instead of 0-based.
-    panClasses[0] = 0;
-    for (int j = 0, nIndex = 1; j < nSpan + 1; j++)
+    if (!bTooMany)
     {
-        if (pabyFlags[j] == 1)
+        // The classes are always padded with a leading 0.  This was for aligning
+        // offsets, or making it a 1-based array instead of 0-based.
+        panClasses[0] = 0;
+        for (int j = 0, nIndex = 1; j < 65536; j++)
```
Comment (elpaso) [inline frmts/raw/lcpdataset.cpp:869]: numeric_limits ?

File: frmts/raw/lcpdataset.cpp (hunk 1 above)
Comment (rouault) [reply frmts/raw/lcpdataset.cpp]: done

File: frmts/raw/lcpdataset.cpp (hunk 2 above)
Comment (rouault) [reply frmts/raw/lcpdataset.cpp]: done

**Comment by rouault** [comment]
The backport to `release/3.6` failed:
```
The process '/usr/bin/git' failed with exit code 1
```
<details>
<summary>stderr</summary>

```
error: could not apply cc18ef1144... LCP: CreateCopy(): fix crash on negative pixel values (fixes #7561)
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

```
</details>

<details>
<summary>stdout</summary>

```
Auto-merging autotest/gdrivers/lcp.py
Auto-merging frmts/raw/lcpdataset.cpp
CONFLICT (content): Merge conflict in frmts/raw/lcpdataset.cpp

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-release/3.6 release/3.6
# Navigate to the new working tree
cd .worktrees/backport-release/3.6
# Create a new branch
git switch --create backport-7564-to-release/3.6
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick cc18ef1144d42dc6989be90a93e4b25bc87a796e
# Push it to GitHub
git push --set-upstream origin backport-7564-to-release/3.6
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-release/3.6
```
Then, create a pull request where the `base` branch is `release/3.6` and the `compare`/`head` branch is `backport-7564-to-release/3.6`.

**Comment by rouault** [comment]
manually backported in 3.6 per e9d98745d34e2995a1ebcb909012676d33dc58d9

**Pull request merged**